trino-aws-proxy-spark4/src/main/java/io/trino/aws/proxy/spark4/PresignAwareS3Client.java

```diff
@@ -0,0 +1,118 @@
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.core.sync.ResponseTransformer;
+import software.amazon.awssdk.services.s3.DelegatingS3Client;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Request;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class PresignAwareS3Client
+        extends DelegatingS3Client
+{
+    private static final Log log = LogFactory.getLog(PresignAwareS3Client.class);
+
+    public PresignAwareS3Client(S3Client delegate)
+    {
+        super(delegate);
+    }
+
+    private Optional<URI> getPresignedUrl(String method, String bucket, String key, Optional<UploadPartRequest> uploadPartRequest)
+    {
+        Optional<String> uploadId = uploadPartRequest.map(UploadPartRequest::uploadId);
+        Optional<Integer> partNumber = uploadPartRequest.map(UploadPartRequest::partNumber);
+
+        log.debug("request.get-presigned-url method=%s bucket=%s key=%s uploadId=%s partNumber=%s".formatted(method, bucket, key, uploadId, partNumber));
+
+        Map<String, List<String>> headers;
+        try {
+            HeadObjectResponse response = headObject(builder -> {
+                builder.bucket(bucket).key(key);
+                partNumber.ifPresent(builder::partNumber);
+                uploadId.ifPresent(id -> builder.overrideConfiguration(c -> c.putRawQueryParameter("uploadId", id)));
+            });
+            headers = response.sdkHttpResponse().headers();
+        }
+        catch (AwsServiceException e) {
+            boolean is404NotFound = (e.statusCode() == 404);
+            boolean isGetOrDelete = method.equalsIgnoreCase("get") || method.equalsIgnoreCase("delete");
+
+            if (is404NotFound && isGetOrDelete) {
+                throw e;
+            }
+
+            // when the bucket isn't found an exception is thrown - but response headers still have pre-signed URLs
+            // given that this not GET/DELETE, we can return the pre-signed URL
+            headers = e.awsErrorDetails().sdkHttpResponse().headers();
+        }
+        catch (RuntimeException e) {
+            log.error("request.get-presigned-url.exception method=%s bucket=%s key=%s uploadId=%s partNumber=%s".formatted(method, bucket, key, uploadId, partNumber), e);
+            throw e;
+        }
+
+        List<String> headerValue = headers.get("X-Trino-Pre-Signed-Url-" + method);
```

Comment:
can we do a `Optional.ofNullable(...).orElse(List.of())` to avoid checking for null values?